### PR TITLE
Consistent `class` and `interface` syntax

### PR DESCRIPTION
**Pull request by josh11b**
Update syntax of `class` and `interface` definitions to be more consistent. Constructs that add names to the class or interface from another definition are always prefixed by the `extend` keyword.

Implements the decisions in:

-   [#995: Generics external impl versus extends](https://github.com/carbon-language/carbon-lang/issues/995),
-   [#1159: adaptor versus adapter may be harder to spell than we'd like](https://github.com/carbon-language/carbon-lang/issues/1159),
-   [#2580: How should Carbon handle conditionally implemented internal interfaces](https://github.com/carbon-language/carbon-lang/issues/2580), and
-   [#2770: Terminology for internal and external implementations](https://github.com/carbon-language/carbon-lang/issues/2770).

### CODE REVIEW

**Comment by josh11b** [comment]
Please pay particular attention to the alternatives, I'm not sure that I have captured everything important we considered in #995 .

**Comment by josh11b** [comment]
I have a draft PR #2866 for updating the syntax in the design docs, if you would like to see how this looks in a number of examples.

File: proposals/p2760.md [hunk 1]
```diff
@@ -0,0 +1,794 @@
...
+-   [#1084: Generics details 9: forward declarations](https://github.com/carbon-language/carbon-lang/pull/1084)
+    allowed forward declaration of implementations, so internal `impl`
+    declarations may appear outside of a class definition, and `external impl`
+    declarations may appear inside.
+-   [#777: Inheritance](https://github.com/carbon-language/carbon-lang/pull/777)
+    defined the syntax for a class to extend a base class.
+
+No proposal so far has defined how forward declarations work for classes. The
+rule used for forward `interface`, `constraint`, and `impl` declarations is that
+declaration part of the definition is everything up to the opening `{` of the
```
Comment (zygoloid) [inline proposals/p2760.md]:
```suggestion
the declaration part of the definition is everything up to the opening `{` of the
```

File: proposals/p2760.md [hunk 2]
```diff
@@ -0,0 +1,794 @@
...
+interface I {
+  // `A`'s interface is incorporated into `I`.
+  extend A;
+
+  // Anything that could be in a `where` clause can
+  // be in a `require` declaration:
+  require A.T == i32;
+
+  // Can combine `extend` with `where`.
+  extend B where B.U == i32;
```
Comment (zygoloid) [inline proposals/p2760.md]:
```suggestion
  extend B where .U == i32;
```
(In `B where ...` we have `.Self: B`, so `.U`, which is `.Self.U`, refers to `.Self.(B.U)`.)

File: proposals/p2760.md [hunk 3]
```diff
@@ -0,0 +1,794 @@
...
+interface C { let V:! Type; }
+interface D { let W:! Type; }
+
+interface I {
+  // `A`'s interface is incorporated into `I`.
+  extend A;
+
+  // Anything that could be in a `where` clause can
+  // be in a `require` declaration:
+  require A.T == i32;
```
Comment (zygoloid) [inline proposals/p2760.md]: I think we don't get an implicit `Self.` here, so this would have to be:

```suggestion
  require Self.(A.T) == i32;
```

Do we get a `.Self` in a `require`? Setting `.Self` to be `Self` might make sense, which would allow this to be shortened to:

```suggestion
  require .(A.T) == i32;
```

We could probably make `A.T == i32` work as a special case, by rewriting it to `Self.(A.T) == i32`. But that seems out of scope for this proposal.

File: proposals/p2760.md [hunk 4]
```diff
@@ -0,0 +1,794 @@
...
+  extend B where B.U == i32;
+
+  // No impact on `I`s interface, but an
+  // implementation must exist.
+  require Self impls C where .V == i32;
+
+  // Again the `where` clause can be located
+  // in a separate `require` declaration.
+  require Self impls D;
+  require D.W == i32;
```
Comment (zygoloid) [inline proposals/p2760.md]:
```suggestion
  require Self.(D.W) == i32;
```
or possibly
```suggestion
  require .(D.W) == i32;
```

File: proposals/p2760.md [hunk 5]
```diff
@@ -0,0 +1,794 @@
...
+Notes:
+
+-   The same change applies to named constraints, and is intended to be used in
+    the future for named predicates used as template constraints.
+-   One syntax for constraints in either `where` clauses or `require`
+    declarations.
+-   Want to open up the syntax to expressing more general constraints.
+-   Syntax for a `require` declaration in an interface or named constraint:
+
+    > `require` _where-constraint-argument_ `;`
```
Comment (zygoloid) [inline proposals/p2760.md]: I remember there being a question about whether rewrite constraints are disallowed here, or whether they're allowed but don't result in rewrites being performed. (Might have been in the context of #2409.) But I don't remember what the conclusion was.

File: proposals/p2760.md [hunk 6]
```diff
@@ -0,0 +1,794 @@
...
+## Rationale
+
+The main reason for the new syntax is consistency and simplification:
+
+-   The use of the `extend` keyword is the consistent way to mark what other
+    entities are consulted during name lookup.
+-   The `class` declaration is simplified by moving more into the definition
+    body.
+-   Making adapters a kind of class removes a kind of top-level declaration,
+    simplification, and matches how base classes are declared, a consistency.
```
Comment (zygoloid) [inline proposals/p2760.md:555]:
```suggestion
    a simplification, and matches how base classes are declared, a consistency.
```
(would have been slightly easier for me to parse, but feel free to ignore this comment).

File: proposals/p2760.md [hunk 7]
```diff
@@ -0,0 +1,794 @@
...
+[#1159: adaptor versus adapter may be harder to spell than we'd like](https://github.com/carbon-language/carbon-lang/issues/1159)
+for the discussion.
+
+We also considered using `adapts` in a class declaration, as in
+`class PlayableSong adapts Song { ... }`, see
+[this comment in #1159](https://github.com/carbon-language/carbon-lang/issues/1159#issuecomment-1316669416).
+This would have also worked, but was not consistent with our resolution of
+[#995: Generics external impl versus extends](https://github.com/carbon-language/carbon-lang/issues/995).
+
+### Use something other syntax for extending adapters
```
Comment (zygoloid) [inline proposals/p2760.md]:
```suggestion
### Use some other syntax for extending adapters
```

File: proposals/p2760.md [hunk 8]
```diff
@@ -0,0 +1,794 @@
...
+
+    However, this felt arbitrary and not compositional.
+
+-   Make the extending and adapting be separate declarations:
+
+    > `extend` _adapted-class_ `;` <br> `adapt` _adapted-class_ `;`
+
+    This felt too repetitive.
+
+-   Make the only way to make an extending adapter is trying to inherit from a
```
Comment (zygoloid) [inline proposals/p2760.md]:
```suggestion
-   Make the only way to make an extending adapter be trying to inherit from a
```

File: proposals/p2760.md [hunk 9]
```diff
@@ -0,0 +1,794 @@
...
+being more convenient to express conditional internal implementation, and wasn't
+seen as attractive once that feature was removed. In the discussion, we
+generally preferred marking additional places to consult in name lookup since
+that was something we expected readers of the code to want to specifically look
+for.
+
+This option was proposed as the third option in the
+[original #995 question](https://github.com/carbon-language/carbon-lang/issues/995#issue-1085174338),
+and was considered in [this
+comment]https://github.com/carbon-language/carbon-lang/issues/995#issuecomment-1006096692).
```
Comment (zygoloid) [inline proposals/p2760.md]:
```suggestion
comment](https://github.com/carbon-language/carbon-lang/issues/995#issuecomment-1006096692).
```

File: proposals/p2760.md [hunk 10]
```diff
@@ -0,0 +1,794 @@
...
+"export" them and leave it up to the mixer class to import them.
+
+We had concerns that there might be name conflicts, but we thought those might
+be handled by some other mechanism. This is being considered in
+question-for-leads issue
+[#2745: Name conflicts beyond inheritance](https://github.com/carbon-language/carbon-lang/issues/2745).
+
+We wanted mixin member names to behave consistently like other class member
+names, and so default to public but can have a `private` modifier to make
+public, following
```
Comment (zygoloid) [inline proposals/p2760.md]:
```suggestion
private, following
```

**Review by zygoloid — COMMENTED** [review]
Thanks, this looks good and aligned with previous discussion and answers to questions-for-leads.

File: proposals/p2760.md (hunk 3 above)
Comment (josh11b) [reply proposals/p2760.md]: What you are suggesting is different from what was agreed in https://github.com/carbon-language/carbon-lang/issues/995#issuecomment-1496781087

File: proposals/p2760.md (hunk 5 above)
Comment (josh11b) [reply proposals/p2760.md]: I would welcome restrictions here -- I would prefer that there were fewer ways of writing equivalent interfaces. In fact, I'm pretty unhappy with the freedom shown in the example above. I think the only requirement is that the syntax matches across `where` and `require` for constraints that are allowed in both.

File: proposals/p2760.md (hunk 3 above)
Comment (zygoloid) [reply proposals/p2760.md]: I think that's an error in @chandlerc's comment, rather than intentionally introducing a new way of naming associated constants with an implicit `Self.`, but let's check once @chandlerc returns from vacation.

Comment (chandlerc) [reply proposals/p2760.md]: FWIW, I hadn't realized or assumed anything specific about `Self` here. It's absence in my comment wasn't a specific intent, just that I happened to not write it.

However, looking at this example, I find both `Self.(A.T)` and `.(A.T)` to be pretty unfortunate.

I'm not sure why `A` is here at all.

Given `extend A;`, why isn't this just `require T == i32`? Shouldn't `T` be in scope after the `extend`?

But maybe it's better to leave a TODO here and revisit this as a follow-up just to make more incremental progress?

File: proposals/p2760.md (hunk 5 above)
Comment (chandlerc) [reply proposals/p2760.md]: Agreed with @josh11b here, I'd prefer to allow the minimal set of things, as long as we don't have *divergence* between `where` and `requires`.

**Review by chandlerc — COMMENTED** [review]
Haven't had time to fully work through, but wanted to at least respond to some of the open threads here.

File: proposals/p2760.md (hunk 3 above)
Comment (josh11b) [reply proposals/p2760.md]: Upon reflection, I think it would be good to restrict this proposal to just `require`...`impls`, and leave everything else for a future proposal.

File: proposals/p2760.md (hunk 5 above)
Comment (josh11b) [reply proposals/p2760.md]: Addressed in https://github.com/carbon-language/carbon-lang/pull/2760/commits/e8f9974245326c1efe4b164650bd6e9477513c13

File: proposals/p2760.md (hunk 3 above)
Comment (josh11b) [reply proposals/p2760.md]: Addressed in https://github.com/carbon-language/carbon-lang/pull/2760/commits/e8f9974245326c1efe4b164650bd6e9477513c13

File: proposals/p2760.md (hunk 4 above)
Comment (josh11b) [reply proposals/p2760.md]: I've removed support for this entirely in https://github.com/carbon-language/carbon-lang/pull/2760/commits/e8f9974245326c1efe4b164650bd6e9477513c13

**Review by zygoloid — APPROVED** [review]
Approving on behalf of leads. This reflects decisions from prior leads issues, with a scope reduction for `require` that @chandlerc and I are both happy with and that seems to generally improve consensus. Thanks!
